Source/JavaScriptCore/runtime/JSGenericTypedArrayViewPrototypeFunctions.h

```diff
@@ -669,6 +669,70 @@ ALWAYS_INLINE EncodedJSValue genericTypedArrayViewPrivateFuncSort(VM& vm, JSGlob
     return JSValue::encode(thisObject);
 }
 
+template<typename ViewClass>
+ALWAYS_INLINE EncodedJSValue genericTypedArrayViewPrivateFuncFromFast(VM& vm, JSGlobalObject* globalObject, CallFrame* callFrame)
+{
+    auto scope = DECLARE_THROW_SCOPE(vm);
+
+    JSValue arrayLike = callFrame->uncheckedArgument(1);
+    JSArrayBufferView* items = jsDynamicCast<JSArrayBufferView*>(arrayLike);
+    if (!items) {
+        // Converting Double or Int32 to BigInt throws an error.
+        if constexpr (ViewClass::TypedArrayStorageType == TypeBigInt64 || ViewClass::TypedArrayStorageType == TypeBigUint64)
+            return JSValue::encode(jsUndefined());
+
+        // TypedArray.from(Array) case.
+        JSArray* array = jsDynamicCast<JSArray*>(arrayLike);
+        if (!array)
+            return JSValue::encode(jsUndefined());
+
+        if (!array->isIteratorProtocolFastAndNonObservable())
+            return JSValue::encode(jsUndefined());
+
+        IndexingType indexingType = array->indexingType() & IndexingShapeMask;
+        if (indexingType != Int32Shape && indexingType != DoubleShape)
+            return JSValue::encode(jsUndefined());
+
+        size_t length = array->length();
+
+        Structure* structure = globalObject->typedArrayStructure(ViewClass::TypedArrayStorageType);
+        ViewClass* result = ViewClass::createUninitialized(globalObject, structure, length);
+        RETURN_IF_EXCEPTION(scope, { });
+
+        if (indexingType == Int32Shape) {
+            for (unsigned i = 0; i < length; i++) {
+                JSValue value = array->butterfly()->contiguous().at(array, i).get();
+                if (LIKELY(!!value))
+                    result->setIndexQuicklyToNativeValue(i, ViewClass::Adaptor::toNativeFromInt32(value.asInt32()));
+                else
+                    result->setIndexQuicklyToNativeValue(i, ViewClass::Adaptor::toNativeFromDouble(PNaN));
```

Comment:
For integer type arrays, the hole maps to `undefined`, which in turn maps to 0.  Going through `toNativeFromDouble` incurs a non-zero cost as it's not constexpr, and when `HAVE(FJCVTZS_INSTRUCTION)`, it routes through inline asm.  Let's implement `toNativeFromUndefined` as constexpr and have it return a 0 for integer type arrays, and PNaN for float type arrays, and use `toNativeFromUndefined` instead.

Reply:
OK, added.